src/main/java/software/amazon/awssdk/crt/http/HttpConnection.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.crt.http;
+
+import software.amazon.awssdk.crt.AsyncCallback;
+import software.amazon.awssdk.crt.CrtResource;
+import software.amazon.awssdk.crt.CrtRuntimeException;
+import software.amazon.awssdk.crt.io.ClientBootstrap;
+import software.amazon.awssdk.crt.io.SocketOptions;
+import software.amazon.awssdk.crt.io.TlsContext;
+
+import java.net.URI;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+
+import static software.amazon.awssdk.crt.CRT.AWS_CRT_SUCCESS;
+
+
+/**
+ * This class wraps aws-c-http to provide the basic HTTP request/response functionality via the AWS Common Runtime.
+ *
+ * HttpConnection represents a single connection to a HTTP service endpoint.
+ *
+ * This class is not thread safe and should not be called from different threads.
+ */
+public class HttpConnection extends CrtResource {
+    private static final String HTTP = "http";
+    private static final String HTTPS = "https";
+
+    private final ClientBootstrap clientBootstrap;
+    private final SocketOptions socketOptions;
+    private final TlsContext tlsContext;
+    private final URI uri;
+    private final HttpConnectionEvents userConnectionCallbacks;
+
+    private ConnectionState connectionState = ConnectionState.DISCONNECTED;
+    private CompletableFuture<Boolean> connectedFuture;
+    private AsyncCallback connectAck;
+    private AsyncCallback disconnectAck;
+
+    public enum ConnectionState {
+        DISCONNECTED, CONNECTING, CONNECTED, SHUTTING_DOWN, SHUT_DOWN
```

Comment:
Probably don't need to track state here in Java.
If the connection is provided via CompletableFuture you could remove the CONNECTING state.

Also, the C implementation will give appropriate errors if its functions are called while it's shutting down or shut down. You can just blithely call into the C-layer without worrying.

Reply:
This is probably echoing the state tracking in the MqttConnection, which is also probably wrong. I thought it was a good idea last year when I wrote it, I'm pretty well convinced of removing it before GA.

Reply:
Removed `ConnectionState` Enum